raiden/network/utils.py

```diff
@@ -10,37 +11,56 @@
 
 
 def get_free_port(
-        bind_address: str = '127.0.0.1',
         initial_port: int = 0,
         socket_kind: SocketKind = SocketKind.SOCK_STREAM,
+        reliable: bool = True,
 ):
     """
     Find an unused TCP port.
-    This should not be used in misson-critical applications - a race condition may occur if
-    someone grabs the port before caller of this function has chance to use it.
+
+    Unless the `reliable` parameter is set to `True` (the default) this is prone to race
```

Comment:
([Helpful TCP state transition reference](http://users.cs.northwestern.edu/~agupta/cs340/project2/TCPIP_State_Transition_Diagram.pdf))

Ok so here's how I understand the problem and solution:

Binding to port 0 makes the OS chose a random 'free' port. From the OS's perspective 'free' means any port that is in `CLOSED` state. 

In the previous implementation we simply bound to the port, grabbed the port number and closed it. That means it's immediately returned into the 'free' pool since its state will be `CLOSED` again.

Since we're running multiple things that require listening ports (and some require more than one port) it appears to me that sometimes the OS will re-use the same port number in successive iterations. 

By forcing the port into `TIME_WAIT` state (by connecting to the socket and then closing the connection) we effectively take it out of the OS's 'free' pool for the duration of that state (usually 60s).

By setting `SO_REUSEADDR` (and relying on the fact that most other 'server' kind of applications do the same) we however allow that port to be reused by another application (i.e. geth, synapse, etc.) while it's still in the `TIME_WAIT` state. AFAIU `SO_REUSEADDR` is independent of process group.

Regarding your question about splitting it. It basically already is. The name of the function is probably bad. It should really be called `get_free_port_generator` since it always returns a generator that will either return ports in increasing order (if you pass an `initial_port > 0`) or just random ports indefinitely.

Reply:
Okay, I see. The race across multiple processes is still there, IOW, this will not prevent another process from using the port. However tix fix the problem with subsequent calls with port `0` (which is really the only thing we care about, since we should not really have a rouge process trying to catch the same port as the test)

Reply:
Exactly